hasher-matcher-actioner/hmalib/models.py

```diff
@@ -16,14 +19,96 @@
+    functionality together.
+
+    Most methods defined are class methods and not static methods. Static
+    methods are unbounded which makes it impossible to override and refer to in
+    sub classes.
+    """
+
+    @classmethod
+    def is_total_countable(cls):
+        """
+        To be declared by sub classes. If true, a counter record is maintained
+        in the same table. Be advised, that the counter is incremented with
+        every 'write' be it a create or update.
+        """
+        return False
+
+    @classmethod
+    def get_total_count_keyname(cls):
+        """
+        Is used to store and retrieve count values. To be implemented by
+        concrete classes alone. Unless is_total_countable() is True, this method
+        need not be implemented.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def get_total_count(cls, table: Table) -> t.Optional[int]:
+        """
+        See `is_total_countable`. If self.is_total_countable is False, return
+        None. Else return the number of writes this table has seen.
+        """
+        if cls.is_total_countable():
+            response = table.get_item(
+                Key={"PK": cls._get_count_pkey(), "SK": cls._get_count_skey()}
+            )
+            return (
+                "Item" in response
+                and int(t.cast(Decimal, response["Item"]["WriteCount"]))
+                or 0
+            )
+        else:
+            return None
+
+    @classmethod
+    def _get_count_pkey(cls) -> str:
+        return "counts"
+
+    @classmethod
+    def _get_count_skey(cls) -> str:
+        return f"count#{cls.get_total_count_keyname()}"
+
+
+class DynamoDBItem(DynamoDBCountMixin):
 
     CONTENT_KEY_PREFIX = "c#"
     SIGNAL_KEY_PREFIX = "s#"
     TYPE_PREFIX = "type#"
 
     def write_to_table(self, table: Table):
-        table.put_item(Item=self.to_dynamodb_item())
+        client: Client = table.meta.client
+        transact_items: t.List[TransactWriteItemTypeDef] = [
+            {
+                "Put": {
+                    "Item": self.to_dynamodb_item(),
+                    "TableName": table.table_name,
+                },
+            },
+        ]
+
+        if self.__class__.is_total_countable():
+            transact_items.append(
+                {
+                    "Update": {
+                        "Key": {
+                            "PK": self._get_count_pkey(),
+                            "SK": self._get_count_skey(),
+                        },
+                        # if_not_exists takes care of initializing the count if it hasn't been done yet.
+                        "UpdateExpression": "SET WriteCount = if_not_exists(WriteCount, :zero) + :one",
+                        "TableName": table.table_name,
+                        "ExpressionAttributeValues": {":one": 1, ":zero": 0},
+                    },
+                }
+            )
+
+        client.transact_write_items(TransactItems=transact_items)
```

Comment:
Any way to sanity check the perf cost then add a comment of its impact or that it is negligible? (if it is non-negligible we may want to refactor in a enable/disable flag in a future PR)

Reply:
Used the script below to do a REPL based perf test.
```python
from hmalib.models import PipelinePDQHashRecord
import boto3
tbl = boto3.resource('dynamodb').Table('dipanjanm-HMADataStore')
p = PipelinePDQHashRecord.get_from_content_id(tbl, 'images/storm-2021-05-03-a2f956db-4eb0-4405-affc-3fe54a262db8-200100.jpg')
p.write_to_table(tbl)

# perf tests
import timeit
timeit.timeit(stmt="p.write_to_table(tbl)", number=200, globals={'p': p, 'tbl': tbl})
```

For 200 writes
| Configuration| Time in Seconds|
|--------------|------------------|
|`[put_item + update_item]` | 14 
|`[put_item]`|13|
|`put_item` using High Level API| 10|

Deferring making this a repeatable test, but it appears the handwavy cost is around 40%. 

Most of it appears to be because of a transaction lock. I'm inferring this from the second configuration taking nearly as much time as the first one.